### PR TITLE
Simplify/update setup of rectangular configurations

### PR DESCRIPTION
**Pull request by apcraig**
- Simplify interaction of kmt_type, ice_data_type, and other namelist to initialize rectangular geometries
- Update diagnostics to avoid undefined values when N/S hemisphere not defined
- Update dynamics to avoid divide by zeros on calculations with no ice
- Add medblock ice_data_type
- Add boxchan tests

### CODE REVIEW

**Comment by apcraig** [comment]
Some test results are here,

https://github.com/CICE-Consortium/Test-Results/wiki/cice_by_hash_forks#e01b2e82295d0753bb7ff2fe235948adae4787a1

Note that the atm forcing changed for most box tests by multiplying forcing by aice (instead of aiu).

File: cicecore/cicedynB/analysis/ice_diagnostics.F90 [hunk 1]
```diff
@@ -389,8 +402,12 @@ subroutine runtime_diags (dt)
       endif
 
       ! maximum ice volume (= mean thickness including open water)
+      hmaxn = c0
+      hmaxs = c0
       hmaxn = global_maxval(vice, distrb_info, lmask_n)
       hmaxs = global_maxval(vice, distrb_info, lmask_s)
+      if (hmaxn < maxval_spval) hmaxn = c0
+      if (hmaxs < maxval_spval) hmaxs = c0
```
Comment (eclare108213) [inline cicecore/cicedynB/analysis/ice_diagnostics.F90:415]: Please explain how this works.  It's not intuitive to me at all!

File: cicecore/cicedynB/dynamics/ice_dyn_evp.F90 [hunk 2]
```diff
@@ -1084,8 +1084,10 @@ subroutine evp (dt)
       do ij = 1, icellu(iblk)
          i = indxui(ij,iblk)
          j = indxuj(ij,iblk)
-         work1(i,j,iblk) = strocnx(i,j,iblk)/aiu(i,j,iblk)
-         work2(i,j,iblk) = strocny(i,j,iblk)/aiu(i,j,iblk)
+         if (aiu(i,j,iblk) /= c0) then
+            work1(i,j,iblk) = strocnx(i,j,iblk)/aiu(i,j,iblk)
+            work2(i,j,iblk) = strocny(i,j,iblk)/aiu(i,j,iblk)
+         endif
```
Comment (eclare108213) [inline cicecore/cicedynB/dynamics/ice_dyn_evp.F90]: A large part of the reason for having the icellu (and similar) loops was efficiency by allowing some vectorization, preventing calculations on cells where things weren't defined, and avoiding lots of if statements like this.  I'd prefer to stick with that philosophy, if possible.

**Review by eclare108213 — COMMENTED** [review]
The box test changes look fine to me. Adding if statements in loops that should be limited to not need them is inconsistent and seems like an unnecessary computational expense.  Is this necessary for the C grid?

**Review by dabail10 — APPROVED** [review]

File: cicecore/cicedynB/analysis/ice_diagnostics.F90 (hunk 1 above)
Comment (apcraig) [reply cicecore/cicedynB/analysis/ice_diagnostics.F90:415]: In general, there may not be a northern or southern region in any grid.  In that case the global_maxval method returns -HUGE.  When that value is written by the formatted writes, it writes "stars".  If the debug flags are turned on, it aborts.  maxval_spval is -HUGE*0.9, so all it's doing is checking whether the global_maxval returned an "undefined" value and then changes it to 0 to write out.  That may not be ideal, but something like it is needed for the formatted write.  The other option is to switch to a different format (901 vs 900 for instance).  Maybe I should do that, but then the log file output will change (formats not values).

File: cicecore/cicedynB/dynamics/ice_dyn_evp.F90 (hunk 2 above)
Comment (apcraig) [reply cicecore/cicedynB/dynamics/ice_dyn_evp.F90]: Yesterday, I tested the case where icetmask=tmask and iceumask=umask.  So computing at all points, not just points with/near ice.  If you do that, there are a few divide by zeros that arise where there is not ice (and we normally wouldn't compute).  So I needed these fixes to run that case.  If we think we don't want to be able to test a larger icetmask/iceumask region that we do now, I can get rid of these again.  I don't think the if should affect performance much anymore.  Happy to undo these changes again.

File: cicecore/cicedynB/analysis/ice_diagnostics.F90 (hunk 1 above)
Comment (eclare108213) [reply cicecore/cicedynB/analysis/ice_diagnostics.F90:415]: ok, thank you

File: cicecore/cicedynB/dynamics/ice_dyn_evp.F90 (hunk 2 above)
Comment (eclare108213) [reply cicecore/cicedynB/dynamics/ice_dyn_evp.F90]: I'd prefer to keep it the way it was, unless there's a situation with the new grids that requires icetmask and iceumask to be redefined with a larger footprint.  This is fine for testing, but then get rid of them (maybe by commenting out with comments indicating what they are useful for).  I'm also willing to be convinced otherwise.

**Review by eclare108213 — APPROVED** [review]

File: cicecore/cicedynB/analysis/ice_diagnostics.F90 (hunk 1 above)
Comment (phil-blain) [reply cicecore/cicedynB/analysis/ice_diagnostics.F90:415]: By the way, that should solve https://github.com/CICE-Consortium/CICE/issues/669, right? that would be great.

@apcraig with the Intel compiler, you can compile with `-nooutput_conversion`  to let it run (and not crash) even in debug mode.

File: cicecore/cicedynB/dynamics/ice_dyn_evp.F90 (hunk 2 above)
Comment (phil-blain) [reply cicecore/cicedynB/dynamics/ice_dyn_evp.F90]: Here might be an actual good use for a CPP macro, maybe ? so that with the right option at case creation, we get the the new behaviour, but without the option, we CPP out the `if/endif` ?

File: cicecore/cicedynB/general/ice_init.F90 [hunk 3]
```diff
@@ -2690,6 +2688,10 @@ subroutine set_state_var (nx_block, ny_block, &
 
       if (trim(ice_ic) == 'default') then
 
+         !---------------------------------------------------------
+         ! ice concentration/thickness
+         !---------------------------------------------------------
+
```
Comment (phil-blain) [inline cicecore/cicedynB/general/ice_init.F90:2694]: these comments are very welcomed, that part of set_state_var can be hard to follow!

File: cicecore/cicedynB/general/ice_init.F90 [hunk 4]
```diff
@@ -2772,7 +2778,20 @@ subroutine set_state_var (nx_block, ny_block, &
             enddo
             enddo
 
-         else if (trim(ice_data_type) == 'smallblock') then
+         elseif (trim(ice_data_type) == 'channel') then
```
Comment (phil-blain) [inline cicecore/cicedynB/general/ice_init.F90:2781]: In general, I think a good way forward (not ncecessarily now) with all these `ice_data_type` values would be to put the code for each in a separate subroutine. It would make `set_state_var` shorter, and easier to follow, I would think.

File: cicecore/cicedynB/infrastructure/ice_grid.F90 [hunk 5]
```diff
@@ -1404,71 +1404,60 @@ subroutine rectgrid
       if (my_task == master_task) then
          work_g1(:,:) = c0      ! initialize hm as land
 
-         if (trim(ew_boundary_type) == 'cyclic') then
+         if (trim(kmt_type) == 'boxislands') then
```
Comment (phil-blain) [inline cicecore/cicedynB/infrastructure/ice_grid.F90:1408]: This cleanup does make the code cleaner, thanks. I found it easier to review with the "Hide whitespace" and "Side-by-side" view, i.e https://github.com/apcraig/CICE/pull/57/files?diff=split&w=1.

File: cicecore/cicedynB/infrastructure/ice_grid.F90 [hunk 6]
```diff
@@ -1404,71 +1404,60 @@ subroutine rectgrid
...
-               endif
+            endif ! > 5x5 grid
 
-            endif ! kmt_type
+         else
 
-         elseif (trim(ew_boundary_type) == 'closed') then
+            call abort_ice(subname//'ERROR: unknown kmt_type '//trim(kmt_type))
 
-            call abort_ice(subname//'ERROR: closed boundaries not available')
```
Comment (phil-blain) [inline cicecore/cicedynB/infrastructure/ice_grid.F90:1469]: OK, we still have an abort in that case in `ice_domain::init_domain_distribution`, so it's OK to remove that one here.

File: cicecore/cicedynB/infrastructure/ice_grid.F90 [hunk 7]
```diff
@@ -1404,71 +1404,60 @@ subroutine rectgrid
...
-
-         elseif (trim(ew_boundary_type) == 'open') then
+            do j = 3,ny_global-2      ! closed top and bottom
+            do i = 1,nx_global        ! open sides
+               work_g1(i,j) = c1    ! NOTE nx_global > 5
+            enddo
+            enddo
 
-            if (trim(kmt_type) == 'boxislands') then
+         elseif (trim(kmt_type) == 'default') then
```
Comment (phil-blain) [inline cicecore/cicedynB/infrastructure/ice_grid.F90:1419]: I was under the impression that we were moving away from using `default` for the name of namelist options, maybe something more descriptive ? can't really think of a good name though...

File: cicecore/cicedynB/analysis/ice_diagnostics.F90 (hunk 1 above)
Comment (apcraig) [reply cicecore/cicedynB/analysis/ice_diagnostics.F90:415]: Yes, should fix #669, glad you put that together.

File: cicecore/cicedynB/dynamics/ice_dyn_evp.F90 (hunk 2 above)
Comment (apcraig) [reply cicecore/cicedynB/dynamics/ice_dyn_evp.F90]: This would be a good place for a CPP, but for now, I'll comment these out.

File: cicecore/cicedynB/infrastructure/ice_grid.F90 (hunk 7 above)
Comment (apcraig) [reply cicecore/cicedynB/infrastructure/ice_grid.F90:1419]: It's a balance of backwards compatibility and cleanup.  I lean towards backward compatibility as the baseline.  I always wonder if people update code and use "old" namelists.  Maybe they don't, but if they do, it'll cause problems.  Maybe we can keep default as an undocumented option but change the documented setting to something else and then support both options as the same.

**Comment by apcraig** [comment]
I removed the mods for avoiding divide by zero in the dynamics.  I think we won't be running dynamics on all points often enough to keep this.

File: cicecore/cicedynB/infrastructure/ice_grid.F90 (hunk 7 above)
Comment (phil-blain) [reply cicecore/cicedynB/infrastructure/ice_grid.F90:1419]: Yes, this has bitten me in the past. I agree then that keeping backward compatibility is better. Thanks!
